### ISSUE TITLE
Invalid context access due to the LLVM's optimization

### ISSUE THREAD

**Issue by mmisono** [iovisor/bcc#2745]
When I try to run the following program, I got an invalid context access error.

```python
#!/usr/bin/env python

import bcc

text = r"""
#include <linux/ptrace.h>

int func(struct pt_regs *ctx) {
  if (ctx->cs & 0x3)
    bpf_trace_printk("hi\n");
  return 0;
}
"""

b = bcc.BPF(text=text, debug=0x8)
b.attach_perf_event(ev_type=bcc.PerfType.SOFTWARE,
                    ev_config=bcc.PerfSWConfig.CPU_CLOCK, fn_name="func",
                    sample_freq=99, cpu=0)
b.trace_print()
```

```
Disassembly of section .bpf.fn.func:
func:
; int func(struct pt_regs *ctx) { // Line  15
   0:   71 11 88 00 00 00 00 00         r1 = *(u8 *)(r1 + 136)
; if (ctx->cs & 0x3) // Line  17
   1:   57 01 00 00 03 00 00 00         r1 &= 3
   2:   15 01 06 00 00 00 00 00         if r1 == 0 goto +6
   3:   b7 01 00 00 68 69 0a 00         r1 = 682344
; ({ char _fmt[] = "hi\n"; bpf_trace_printk_(_fmt, sizeof(_fmt)); }); // Line  18
   4:   63 1a fc ff 00 00 00 00         *(u32 *)(r10 - 4) = r1
   5:   bf a1 00 00 00 00 00 00         r1 = r10
   6:   07 01 00 00 fc ff ff ff         r1 += -4
   7:   b7 02 00 00 04 00 00 00         r2 = 4
   8:   85 00 00 00 06 00 00 00         call 6
; return 0; // Line  19
   9:   b7 00 00 00 00 00 00 00         r0 = 0
  10:   95 00 00 00 00 00 00 00         exit

bpf: Failed to load program: Permission denied
0: (71) r1 = *(u8 *)(r1 +136)
invalid bpf_context access off=136 size=1
processed 1 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
```

The error reason is LLVM generates a byte load instruction (`r1 = *(u8 *)(r1 + 136)`) but [it is not allowed](https://github.com/torvalds/linux/blob/6992ca0dd017ebaa2bf8e9dcc49f1c3b7033b082/kernel/trace/bpf_trace.c#L1284-L1285) for some reason.

If I mark the argument of the function as volatile, then the program works as intended.

```c
int func(volatile struct pt_regs *ctx) {
  if (ctx->cs & 0x3)
    bpf_trace_printk("hi\n");
  return 0;
}
```

```
Disassembly of section .bpf.fn.func:
func:
; int func(volatile struct pt_regs *ctx) { // Line  15
   0:   79 11 88 00 00 00 00 00         r1 = *(u64 *)(r1 + 136)
[...]
```

However, I'm not sure this is the right way to cope with this. I'm wondering if bcc'c rewriter help to solve this.

(I originally found this issue when [using bpftrace](https://github.com/iovisor/bpftrace/pull/1104/commits/eac74051c72d6aa2723491ad3e446c052407d929))

**Comment by yonghong-song** [comment]
Using the volatile to workaround the issue is the right way to do at this point. The issue is fixed in llvm trunk (llvm 11.0): https://github.com/llvm/llvm-project/commit/d96c1bbaa03574daf759e5e9a6c75047c5e3af64#diff-6e02af2c61254c488f583a59982eb684

You can also workaround the issue with `(*(volatile unsigned long *)&(ctx->cs)) & 0x3`, I think.

**Comment by mmisono** [comment]
Thanks for the information. It is very helpful. It seems I hit the issue that you fixed very recently.
